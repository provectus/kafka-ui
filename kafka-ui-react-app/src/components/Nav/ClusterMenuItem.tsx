import React, { PropsWithChildren } from 'react';

import * as S from './Nav.styled';

export interface ClusterMenuItemProps {
  to: string;
  title?: string;
  isTopLevel?: boolean;
}

const ClusterMenuItem: React.FC<PropsWithChildren<ClusterMenuItemProps>> = (
  props
) => {
  const { to, title, children, isTopLevel } = props;

  if (to) {
    return (
      <S.ListItem $isTopLevel={isTopLevel}>
        <S.Link to={to} title={title}>
          {title}
        </S.Link>
        {children}
      </S.ListItem>
    );
  }

  return <S.ListItem {...props} />;
};

export default ClusterMenuItem;
