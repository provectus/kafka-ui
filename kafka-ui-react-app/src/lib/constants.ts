import { SelectOption } from 'components/common/Select/Select';
import { ConfigurationParameters } from 'generated-sources';

declare global {
  interface Window {
    basePath: string;
  }
}

export const BASE_PARAMS: ConfigurationParameters = {
  basePath: window.basePath || '',
  credentials: 'include',
  headers: {
    'Content-Type': 'application/json',
  },
};

export const TOPIC_NAME_VALIDATION_PATTERN = /^[a-zA-Z0-9._-]+$/;
export const SCHEMA_NAME_VALIDATION_PATTERN = /^[.,A-Za-z0-9_/-]+$/;

export const TOPIC_CUSTOM_PARAMS_PREFIX = 'customParams';
export const TOPIC_CUSTOM_PARAMS: Record<string, string> = {
  'compression.type': 'producer',
  'leader.replication.throttled.replicas': '',
  'message.downconversion.enable': 'true',
  'segment.jitter.ms': '0',
  'flush.ms': '9223372036854775807',
  'follower.replication.throttled.replicas': '',
  'segment.bytes': '1073741824',
  'flush.messages': '9223372036854775807',
  'message.format.version': '2.3-IV1',
  'file.delete.delay.ms': '60000',
  'max.compaction.lag.ms': '9223372036854775807',
  'min.compaction.lag.ms': '0',
  'message.timestamp.type': 'CreateTime',
  preallocate: 'false',
  'min.cleanable.dirty.ratio': '0.5',
  'index.interval.bytes': '4096',
  'unclean.leader.election.enable': 'true',
  'retention.bytes': '-1',
  'delete.retention.ms': '86400000',
  'segment.ms': '604800000',
  'message.timestamp.difference.max.ms': '9223372036854775807',
  'segment.index.bytes': '10485760',
};

export const MILLISECONDS_IN_WEEK = 604_800_000;
export const MILLISECONDS_IN_DAY = 86_400_000;
export const MILLISECONDS_IN_SECOND = 1_000;

export const NOT_SET = -1;
export const BYTES_IN_GB = 1_073_741_824;
export const BUILD_VERSION_PATTERN = /v\d.\d.\d/;

export const PER_PAGE = 25;
export const MESSAGES_PER_PAGE = '100';

export const GIT_REPO_LINK = 'https://github.com/provectus/kafka-ui';
export const GIT_REPO_LATEST_RELEASE_LINK =
  'https://api.github.com/repos/provectus/kafka-ui/releases/latest';

export const LOCAL_STORAGE_KEY_PREFIX = 'kafka-ui';

export enum AsyncRequestStatus {
  initial = 'initial',
  pending = 'pending',
  fulfilled = 'fulfilled',
  rejected = 'rejected',
}

export const QUERY_REFETCH_OFF_OPTIONS = {
  refetchOnMount: false,
  refetchOnWindowFocus: false,
  refetchIntervalInBackground: false,
};

// Cluster Form Constants
export const AUTH_OPTIONS: SelectOption[] = [
  { value: 'SASL/JAAS', label: 'SASL/JAAS' },
  { value: 'SASL/GSSAPI', label: 'SASL/GSSAPI' },
  { value: 'SASL/OAUTHBEARER', label: 'SASL/OAUTHBEARER' },
  { value: 'SASL/PLAIN', label: 'SASL/PLAIN' },
  { value: 'SASL/SCRAM-256', label: 'SASL/SCRAM-256' },
  { value: 'SASL/SCRAM-512', label: 'SASL/SCRAM-512' },
  { value: 'Delegation tokens', label: 'Delegation tokens' },
  { value: 'SASL/LDAP', label: 'SASL/LDAP' },
  { value: 'SASL/AWS IAM', label: 'SASL/AWS IAM' },
  { value: 'mTLS', label: 'mTLS' },
];

export const SECURITY_PROTOCOL_OPTIONS: SelectOption[] = [
  { value: 'SASL_SSL', label: 'SASL_SSL' },
  { value: 'SASL_PLAINTEXT', label: 'SASL_PLAINTEXT' },
];
export const METRICS_OPTIONS: SelectOption[] = [
  { value: 'JMX', label: 'JMX' },
  { value: 'PROMETHEUS', label: 'PROMETHEUS' },
];
