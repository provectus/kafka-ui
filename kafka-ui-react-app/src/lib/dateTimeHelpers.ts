export const formatTimestamp = (
  timestamp?: number | string | Date,
  format: Intl.DateTimeFormatOptions = { hour12: false }
): string => {
  if (!timestamp) {
    return '';
  }

  // empty array gets the default one from the browser
  const date = new Date(timestamp);

  // invalid date
  if (Number.isNaN(date.getTime())) {
    return '';
  }

  // browser support
  const language = navigator.language || navigator.languages[0];
  return date.toLocaleString(language || [], format);
};

export const formatMilliseconds = (input = 0) => {
  const milliseconds = Math.max(input || 0, 0);

  const seconds = Math.floor(milliseconds / 1000);
  const minutes = Math.floor(seconds / 60);
  const hours = Math.floor(minutes / 60);

  if (hours > 0) {
    return `${hours}h ${minutes % 60}m`;
  }

  if (minutes > 0) {
    return `${minutes}m ${seconds % 60}s`;
  }

  if (seconds > 0) {
    return `${seconds}s`;
  }

  return `${milliseconds}ms`;
};

export const passedTime = (value: number) => (value < 10 ? `0${value}` : value);

export const calculateTimer = (startedAt: number) => {
  const nowDate = new Date();
  const now = nowDate.getTime();
  const newDate = now - startedAt;
  const minutes = nowDate.getMinutes();
  const second = nowDate.getSeconds();

  return newDate > 0 ? `${passedTime(minutes)}:${passedTime(second)}` : '00:00';
};
