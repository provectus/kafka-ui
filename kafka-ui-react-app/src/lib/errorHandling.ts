import { ServerResponse } from 'redux/interfaces';

export const getResponse = async (
  response: Response
): Promise<ServerResponse> => {
  let body;
  try {
    body = await response.json();
  } catch (e) {
    // do nothing;
  }
  return {
    status: response.status,
    statusText: response.statusText,
    url: response.url,
    message: body?.message,
  };
};
