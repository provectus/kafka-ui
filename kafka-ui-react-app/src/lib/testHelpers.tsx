import React, { PropsWithChildren, ReactElement, useMemo } from 'react';
import {
  MemoryRouter,
  MemoryRouterProps,
  Route,
  Routes,
} from 'react-router-dom';
import fetchMock from 'fetch-mock';
import { Provider } from 'react-redux';
import { ThemeProvider } from 'styled-components';
import { theme } from 'theme/theme';
import {
  render,
  renderHook,
  RenderOptions,
  waitFor,
} from '@testing-library/react';
import { AnyAction, Store } from 'redux';
import { RootState } from 'redux/interfaces';
import { configureStore } from '@reduxjs/toolkit';
import rootReducer from 'redux/reducers';
import {
  QueryClient,
  QueryClientProvider,
  UseQueryResult,
} from '@tanstack/react-query';
import { ConfirmContextProvider } from 'components/contexts/ConfirmContext';
import ConfirmationModal from 'components/common/ConfirmationModal/ConfirmationModal';
import { GlobalSettingsContext } from 'components/contexts/GlobalSettingsContext';
import { UserInfoRolesAccessContext } from 'components/contexts/UserInfoRolesAccessContext';

import { RolesType, modifyRolesData } from './permissions';

interface CustomRenderOptions extends Omit<RenderOptions, 'wrapper'> {
  preloadedState?: Partial<RootState>;
  store?: Store<Partial<RootState>, AnyAction>;
  initialEntries?: MemoryRouterProps['initialEntries'];
  userInfo?: {
    roles?: RolesType;
    rbacFlag: boolean;
  };
}

interface WithRouteProps {
  children: React.ReactNode;
  path: string;
}

export const expectQueryWorks = async (
  mock: fetchMock.FetchMockStatic,
  result: { current: UseQueryResult<unknown, unknown> }
) => {
  await waitFor(() => expect(result.current.isFetched).toBeTruthy());
  expect(mock.calls()).toHaveLength(1);
  expect(result.current.data).toBeDefined();
};

export const WithRoute: React.FC<WithRouteProps> = ({ children, path }) => {
  return (
    <Routes>
      <Route path={path} element={children} />
    </Routes>
  );
};

export const TestQueryClientProvider: React.FC<PropsWithChildren<unknown>> = ({
  children,
}) => {
  // use new QueryClient instance for each test run to avoid issues with cache
  const queryClient = new QueryClient({
    defaultOptions: { queries: { retry: false } },
  });
  return (
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  );
};

/**
 * @description it will create a UserInfo Provider that will actually
 * disable the rbacFlag , to user if you can pass it as an argument
 * */
const TestUserInfoProvider: React.FC<
  PropsWithChildren<{ data?: { roles?: RolesType; rbacFlag: boolean } }>
> = ({ children, data }) => {
  const contextValue = useMemo(() => {
    const roles = modifyRolesData(data?.roles);

    return {
      username: 'test',
      rbacFlag: !!(typeof data?.rbacFlag === 'undefined'
        ? false
        : data?.rbacFlag),
      roles,
    };
  }, [data]);

  return (
    <UserInfoRolesAccessContext.Provider value={contextValue}>
      {children}
    </UserInfoRolesAccessContext.Provider>
  );
};

const customRender = (
  ui: ReactElement,
  {
    preloadedState,
    store = configureStore<RootState>({
      reducer: rootReducer,
      preloadedState,
    }),
    initialEntries,
    userInfo,
    ...renderOptions
  }: CustomRenderOptions = {}
) => {
  // overrides @testing-library/react render.
  const AllTheProviders: React.FC<PropsWithChildren<unknown>> = ({
    children,
  }) => (
    <TestQueryClientProvider>
      <GlobalSettingsContext.Provider value={{ hasDynamicConfig: false }}>
        <ThemeProvider theme={theme}>
          <TestUserInfoProvider data={userInfo}>
            <ConfirmContextProvider>
              <Provider store={store}>
                <MemoryRouter initialEntries={initialEntries}>
                  <div>
                    {children}
                    <ConfirmationModal />
                  </div>
                </MemoryRouter>
              </Provider>
            </ConfirmContextProvider>
          </TestUserInfoProvider>
        </ThemeProvider>
      </GlobalSettingsContext.Provider>
    </TestQueryClientProvider>
  );
  return render(ui, { wrapper: AllTheProviders, ...renderOptions });
};

const customRenderHook = (hook: () => UseQueryResult<unknown, unknown>) =>
  renderHook(hook, { wrapper: TestQueryClientProvider });

export { customRender as render, customRenderHook as renderQueryHook };

export class EventSourceMock {
  url: string;

  close: () => void;

  open: () => void;

  error: () => void;

  onmessage: () => void;

  constructor(url: string) {
    this.url = url;
    this.open = jest.fn();
    this.error = jest.fn();
    this.onmessage = jest.fn();
    this.close = jest.fn();
  }
}
