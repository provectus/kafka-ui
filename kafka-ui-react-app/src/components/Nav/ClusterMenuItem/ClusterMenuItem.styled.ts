import styled from 'styled-components';

import { MenuItemProps } from './ClusterMenuItem';

const StyledMenuItem = styled.li<MenuItemProps>`
  font-size: 14px;
  font-weight: ${(props) => (props.isTopLevel ? 500 : 'normal')};
  height: 32px;
  display: flex;
  user-select: none;
  & a {
    width: 100%;
    padding: 0.5em 0.75em;
    cursor: pointer;
    text-decoration: none;
    margin: 0px 0px;
    background-color: ${(props) =>
      props.theme.menuStyles.backgroundColor.normal};
    color: ${(props) => props.theme.menuStyles.color.normal};

    &:hover {
      background-color: ${(props) =>
        props.theme.menuStyles.backgroundColor.hover};
      color: ${(props) => props.theme.menuStyles.color.hover};
    }
    &.is-active {
      background-color: ${(props) =>
        props.theme.menuStyles.backgroundColor.active};
      color: ${(props) => props.theme.menuStyles.color.active};
    }
  }
`;

StyledMenuItem.displayName = 'StyledMenuItem';

export default StyledMenuItem;
