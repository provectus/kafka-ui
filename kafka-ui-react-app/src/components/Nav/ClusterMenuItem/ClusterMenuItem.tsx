import { StyledMenuItem } from 'components/Nav/ClusterMenuItem/ClusterMenuItem.styled';
import React from 'react';
import { NavLink } from 'react-router-dom';

export interface MenuItemProps {
  to: string;
  activeClassName?: string;
  title?: string;
  isInverted?: boolean;
  exact?: boolean;
  isTopLevel?: boolean;
  isActive?: (match: unknown, location: Location) => boolean;
}

const ClusterMenuItem: React.FC<MenuItemProps> = (props) => {
  const {
    to,
    activeClassName,
    title,
    children,
    isActive,
    exact,
    isTopLevel,
    ...rest
  } = props;

  if (to) {
    return (
      <StyledMenuItem to={to} isTopLevel={isTopLevel}>
        <NavLink
          to={to}
          activeClassName={activeClassName}
          title={title}
          exact={exact}
          {...rest}
        >
          {title}
        </NavLink>
        {children}
      </StyledMenuItem>
    );
  }

  return <StyledMenuItem {...props} />;
};

export default ClusterMenuItem;
