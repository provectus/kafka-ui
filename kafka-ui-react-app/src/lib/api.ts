import {
  KsqlApi,
  TopicsApi,
  SchemasApi,
  BrokersApi,
  MessagesApi,
  ClustersApi,
  Configuration,
  KafkaConnectApi,
  ConsumerGroupsApi,
  AuthorizationApi,
  ApplicationConfigApi,
} from 'generated-sources';
import { BASE_PARAMS } from 'lib/constants';

const apiClientConf = new Configuration(BASE_PARAMS);

export const ksqlDbApiClient = new KsqlApi(apiClientConf);
export const topicsApiClient = new TopicsApi(apiClientConf);
export const brokersApiClient = new BrokersApi(apiClientConf);
export const schemasApiClient = new SchemasApi(apiClientConf);
export const messagesApiClient = new MessagesApi(apiClientConf);
export const clustersApiClient = new ClustersApi(apiClientConf);
export const kafkaConnectApiClient = new KafkaConnectApi(apiClientConf);
export const consumerGroupsApiClient = new ConsumerGroupsApi(apiClientConf);
export const authApiClient = new AuthorizationApi(apiClientConf);
export const appConfigApiClient = new ApplicationConfigApi(apiClientConf);
