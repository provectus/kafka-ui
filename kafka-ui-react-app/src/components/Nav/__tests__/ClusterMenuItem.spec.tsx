import React from 'react';
import ClusterMenuItem, {
  ClusterMenuItemProps,
} from 'components/Nav/ClusterMenuItem';
import { screen } from '@testing-library/react';
import { render } from 'lib/testHelpers';

describe('ClusterMenuItem', () => {
  const setupComponent = (props: Partial<ClusterMenuItemProps> = {}) => (
    <ul>
      <ClusterMenuItem to="/test" {...props} />
    </ul>
  );

  const getMenuItem = () => screen.getByRole('menuitem');
  const getLink = () => screen.queryByRole('link');

  it('renders component with correct title', () => {
    const testTitle = 'My Test Title';
    render(setupComponent({ title: testTitle }));
    expect(screen.getByText(testTitle)).toBeInTheDocument();
  });

  it('renders top level component with correct styles', () => {
    render(setupComponent({ isTopLevel: true }));
    expect(getMenuItem()).toHaveStyle({ fontWeight: '500' });
  });

  it('renders non-top level component with correct styles', () => {
    render(setupComponent({ isTopLevel: false }));
    expect(getMenuItem()).toHaveStyle({ fontWeight: 'normal' });
  });

  it('renders list item with link inside', () => {
    render(setupComponent({ to: '/my-cluster' }));
    expect(getMenuItem()).toBeInTheDocument();
    expect(getLink()).toBeInTheDocument();
  });

  it('renders list item without link inside', () => {
    render(setupComponent({ to: '' }));
    expect(getMenuItem()).toBeInTheDocument();
    expect(getLink()).not.toBeInTheDocument();
  });

  it('renders list item with children', () => {
    render(
      <ul>
        <ClusterMenuItem to="/test">Test Text Box</ClusterMenuItem>
      </ul>
    );
    expect(getMenuItem()).toBeInTheDocument();
    expect(getLink()).toBeInTheDocument();
    expect(screen.getByText('Test Text Box')).toBeInTheDocument();
  });
});
