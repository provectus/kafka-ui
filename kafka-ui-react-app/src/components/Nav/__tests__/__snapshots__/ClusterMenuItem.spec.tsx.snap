// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClusterMenuItem matches the snapshot 1`] = `
.c0 {
  font-size: 14px;
  font-weight: normal;
  height: 32px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c0 a {
  width: 100%;
  padding: 0.5em 0.75em;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin: 0px 0px;
  background-color: #FFFFFF;
  color: #73848C;
}

.c0 a:hover {
  background-color: #F1F2F3;
  color: #73848C;
}

.c0 a.is-active {
  background-color: #E3E6E8;
  color: #171A1C;
}

<StaticRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
    staticContext={Object {}}
  >
    <ClusterMenuItem
      to="/test"
    >
      <styled.li
        to="/test"
      >
        <li
          className="c0"
          to="/test"
        >
          <NavLink
            to="/test"
          >
            <Link
              aria-current={null}
              to={
                Object {
                  "hash": "",
                  "pathname": "/test",
                  "search": "",
                  "state": null,
                }
              }
            >
              <LinkAnchor
                aria-current={null}
                href="/test"
                navigate={[Function]}
              >
                <a
                  aria-current={null}
                  href="/test"
                  onClick={[Function]}
                />
              </LinkAnchor>
            </Link>
          </NavLink>
        </li>
      </styled.li>
    </ClusterMenuItem>
  </Router>
</StaticRouter>
`;
